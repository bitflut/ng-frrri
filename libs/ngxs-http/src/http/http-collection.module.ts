import { CommonModule } from '@angular/common';
import { ModuleWithProviders, NgModule } from '@angular/core';
import { HttpCollectionService } from './http-collection.service';
import { PaginatedHttpCollectionService } from './http-paginated-collection.service';

@NgModule({
    imports: [CommonModule],
})
export class HttpCollectionModule {

    constructor(
    ) { }

    static forRoot(): ModuleWithProviders {
        return {
            ngModule: HttpCollectionModule,
            providers: [
                HttpCollectionService,
                PaginatedHttpCollectionService,
            ],
        };
    }
}
