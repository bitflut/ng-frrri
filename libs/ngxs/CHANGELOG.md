# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.0.3-alpha.0](https://github.com/bitflut/ng-frrri/compare/@ng-frrri/ngxs@0.0.2...@ng-frrri/ngxs@0.0.3-alpha.0) (2020-05-26)


### Features

* build scripts ([8294ed4](https://github.com/bitflut/ng-frrri/commit/8294ed42f94f174d968dc4074f03e3d1347b5c81))





## [0.0.2](https://github.com/bitflut/ng-frrri/compare/@ng-frrri/ngxs@0.1.0...@ng-frrri/ngxs@0.0.2) (2020-05-26)

**Note:** Version bump only for package @ng-frrri/ngxs





# 0.1.0 (2020-05-26)



## 0.0.1 (2020-05-26)


### Features

* [@ng-frrri](https://github.com/ng-frrri) ([736ca8c](https://github.com/bitflut/ng-frrri/commit/736ca8cda25f2c01e3ae1ca834ed90f528578419))
* deactivated platform ([#2](https://github.com/bitflut/ng-frrri/issues/2)) ([d3efdfc](https://github.com/bitflut/ng-frrri/commit/d3efdfcc44d5c31558220ecdfc7741e28f2f1a03))
* update libs package.json ([9aa2302](https://github.com/bitflut/ng-frrri/commit/9aa23023dc203a420e23a09e2f49dac34b5304cb))
