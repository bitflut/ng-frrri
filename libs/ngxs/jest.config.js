module.exports = {
  name: 'ngxs',
  preset: '../../jest.config.js',
  coverageDirectory: '../../coverage/libs/ngxs',
  snapshotSerializers: [
    'jest-preset-angular/build/AngularNoNgAttributesSnapshotSerializer.js',
    'jest-preset-angular/build/AngularSnapshotSerializer.js',
    'jest-preset-angular/build/HTMLCommentSerializer.js'
  ]
};
