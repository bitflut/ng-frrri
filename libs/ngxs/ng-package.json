{
  "$schema": "../../node_modules/ng-packagr/package.schema.json",
  "lib": {
    "entryFile": "src/public_api.ts",
    "umdModuleIds": {
      "@ngxs-labs/data/decorators": "@ngxs-labs/data/decorators",
      "@ngxs-labs/data/repositories": "@ngxs-labs/data/repositories",
      "@ngxs-labs/data/internals": "@ngxs-labs/data/internals",
      "@ngxs-labs/data/utils": "@ngxs-labs/data/utils",
      "@ngxs/store": "@ngxs/store",
      "@ngxs/store/internals": "@ngxs/store/internals",
      "@ng-frrri/router-middleware": "@ng-frrri/router-middleware",
      "@ng-frrri/router-middleware/internal": "@ng-frrri/router-middleware/internal",
      "@ng-frrri/router-middleware/operators": "@ng-frrri/router-middleware/operators",
      "parse-link-header": "parse-link-header",
      "lodash": "lodash"
    }
  },
  "dest": "dist"
}
