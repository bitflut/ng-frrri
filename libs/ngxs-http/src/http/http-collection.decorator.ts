import { StateRepository } from '@ngxs-labs/data/decorators';
import { createEntityCollections } from '@ngxs-labs/data/utils';
import { State } from '@ngxs/store';
import { StateClass } from '@ngxs/store/internals';
import { CollectionReducer, CollectionOptions, CollectionStateOptions } from '@ng-frrri/ngxs';
import { HttpCollectionService } from './http-collection.service';

export function HttpCollection<T = CollectionReducer>(options: CollectionOptions<T>) {
    options = {
        ...options,
        baseUrl: options.baseUrl ?? '/api',
        endpoint: options.endpoint ?? options.name.toString(),
        defaults: {
            ...createEntityCollections(),
            active: undefined,
            loaded: false,
            loading: {},
            error: {},
            ...options.defaults,
        },
    } as CollectionOptions<T>;

    const stateFn = State(options);
    const stateRepositoryFn = StateRepository();
    return function (target: StateClass) {
        stateFn(target);
        stateRepositoryFn(target);
        target.prototype.serviceToken = HttpCollectionService;

        const stateOptions = {
            requestOptions: options.requestOptions,
            endpoint: options.endpoint!,
            baseUrl: options.baseUrl!,
        } as CollectionStateOptions;
        if (options.idKey) {
            stateOptions.idKey = options.idKey!;
        }
        target.prototype.stateOptions = stateOptions;
    };
}
