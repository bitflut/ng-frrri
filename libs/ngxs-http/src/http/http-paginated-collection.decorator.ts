import { CollectionReducer } from '@ng-frrri/ngxs';
import { HttpCollection } from './http-collection.decorator';
import { StateClass } from '@ngxs/store/internals';
import { PaginatedCollectionOptions } from '@ng-frrri/ngxs/pagination';
import { PaginatedHttpCollectionService } from './http-paginated-collection.service';

export function PaginatedHttpCollection<T = CollectionReducer>(options: PaginatedCollectionOptions<T>) {
    options = {
        ...options,
        defaults: {
            next: undefined,
            ...options.defaults,
        },
    };

    const crudCollectionFn = HttpCollection(options);
    return function (target: StateClass) {
        target.prototype.pageSize = options.size;
        target.prototype.paginatedServiceToken = PaginatedHttpCollectionService;
        crudCollectionFn(target);
    };
}
