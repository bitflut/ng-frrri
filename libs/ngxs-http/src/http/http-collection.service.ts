import { HttpClient } from '@angular/common/http';
import { Injectable, Injector } from '@angular/core';
import { GetManyOptions } from '@ng-frrri/ngxs/internal';
import { Observable } from 'rxjs';
import { CollectionService, CollectionStateOptions } from '@ng-frrri/ngxs';

@Injectable({
    providedIn: 'root',
})
export class HttpCollectionService<V = any, IdType = any> implements CollectionService<V, IdType> {

    protected http = this.injector.get(HttpClient);

    constructor(protected injector: Injector) { }

    getOne(stateOptions: CollectionStateOptions, id: IdType) {
        const url = stateOptions.requestOptions.resourceUrlFactory(id);
        return this.http.get<V>(url);
    }

    getMany(stateOptions: CollectionStateOptions, options: GetManyOptions = {}): Observable<V[]> {
        const url = stateOptions.requestOptions.collectionUrlFactory();
        return this.http.get<V[]>(url, { params: options.params });
    }

    patchOne(stateOptions: CollectionStateOptions, id: IdType, changes: { [key: string]: Partial<V> }) {
        const url = stateOptions.requestOptions.resourceUrlFactory(id);
        return this.http.patch<V>(url, changes);
    }

    putOne(stateOptions: CollectionStateOptions, id: IdType, changes: { [key: string]: Partial<V> }): Observable<V> {
        const url = stateOptions.requestOptions.collectionUrlFactory();
        return this.http.put<V>(url, changes);
    }

    deleteOne(stateOptions: CollectionStateOptions, id: IdType) {
        const url = stateOptions.requestOptions.resourceUrlFactory(id);
        return this.http.delete<void>(url);
    }

    postOne(stateOptions: CollectionStateOptions, body: Partial<V>): Observable<V> {
        const url = stateOptions.requestOptions.collectionUrlFactory();
        return this.http.post<V>(
            url,
            body,
        );
    }

}
