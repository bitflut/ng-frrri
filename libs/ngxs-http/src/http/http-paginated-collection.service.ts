import { HttpClient } from '@angular/common/http';
import { Injectable, Injector } from '@angular/core';
import { GetManyOptions } from '@ng-frrri/ngxs/internal';
import { EMPTY, Observable } from 'rxjs';
import { expand, reduce } from 'rxjs/operators';
import { CollectionStateOptions } from '@ng-frrri/ngxs';
import { PaginatedCollectionService, Paginated } from '@ng-frrri/ngxs/pagination';

@Injectable({
    providedIn: 'root',
})
export class PaginatedHttpCollectionService<V = any> implements PaginatedCollectionService<V> {

    protected http = this.injector.get(HttpClient);

    constructor(protected injector: Injector) { }

    getMany(stateOptions: CollectionStateOptions, options: GetManyOptions & { size?: number } = {}) {
        const url = stateOptions.requestOptions.collectionUrlFactory();
        return this.http.get<Paginated<V[]>>(url, { params: options.params });
    }

    getAll(stateOptions: CollectionStateOptions, options: GetManyOptions & { size?: number } = {}): Observable<V[]> {
        const url = stateOptions.requestOptions.collectionUrlFactory();
        return this.getManyWithUrl(url, options).pipe(
            expand(response => {
                let next = response.pagination?.next;
                if (!next) { return EMPTY; }

                const isUrlWith = next.startsWith('http');
                if (!isUrlWith) {
                    const baseUrl = url.match(/^(https?:\/\/[^\/]+)\//)[1];
                    next = `${baseUrl}${next}`;
                }

                return this.getManyWithUrl(next);
            }),
            reduce((acc, res) => {
                return [
                    ...acc,
                    ...res.pagination.data,
                ];
            }, [] as V[]),
        );
    }

    getNext(url: any) {
        return this.http.get<Paginated<V[]>>(url);
    }

    private getManyWithUrl(url: string, options: GetManyOptions & { size?: number } = {}): Observable<Paginated<V[]>> {
        return this.http.get<Paginated<V[]>>(url, { params: options.params });
    }

}
