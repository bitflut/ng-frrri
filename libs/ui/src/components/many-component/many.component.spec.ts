import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { Injectable } from '@angular/core';
import { ComponentFixture, inject, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { CollectionState, CrudEntities, CrudEntitiesState, NgxsMiddlewareModule } from '@ng-frrri/ngxs';
import { PaginatedCollectionState, PaginationInterceptor } from '@ng-frrri/ngxs/pagination';
import { NgxsDataPluginModule } from '@ngxs-labs/data';
import { NgxsModule } from '@ngxs/store';
import { MockRender } from 'ng-mocks';
import { Subject } from 'rxjs';
import { take, tap } from 'rxjs/operators';
import { TestPaginatedCollection, TestPaginatedCrudService } from '../../../../ngxs/pagination/src/paginated-collection-state/paginated-collection.state.spec';
import { TestCrudCollection, TestCrudCollectionService } from '../../../../ngxs/src/libs/collection-state/collection.state.spec';
import { ManyComponent } from './many.component';
import { ManyUiModule } from './many.module';

interface Post {
    userId: number;
    id: number;
    body: string;
    title: string;
}

@TestPaginatedCollection({
    baseUrl: 'https://jsonplaceholder.typicode.com',
    name: 'posts',
})
@Injectable()
class PostsEntitiesState extends PaginatedCollectionState<Post, number> { }

interface Comment {
    postId: number;
    id: number;
    name: string;
    body: string;
    email: string;
}

@TestCrudCollection({
    name: 'comments',
    baseUrl: 'https://jsonplaceholder.typicode.com',
})
@Injectable()
class CommentsEntitiesState extends CollectionState<Comment, number> { }

@CrudEntities({
    name: 'cache',
    defaults: {},
    children: [PostsEntitiesState, CommentsEntitiesState],
})
@Injectable()
class EntityCrudEntitiesState extends CrudEntitiesState<any> { }

const page1Data = {
    body: [{
        userId: 1,
        id: 1,
        body: 'Hello World',
        title: 'testing Angular',
    }, {
        userId: 2,
        id: 2,
        body: 'Hello World2',
        title: 'testing Angular2',
    }],
    headers: {
        Link: '</api/v1/admin/posts/?sort=name&pageSize=20&next%5B%24or%5D%5B0%5D%5Bname%5D%5B%24gt%5D=Forum%20Mannheim&next%5B%24or%5D%5B1%5D%5Bname%5D%5B%24gte%5D=Forum%20Mannheim&next%5B%24or%5D%5B1%5D%5B_id%5D%5B%24gt%5D=5e4d2901871f09a8d0fc4bef>; rel="next"',
    },
};

const page2Data = {
    body: [{
        userId: 1,
        id: 3,
        body: 'Hello World3',
        title: 'testing Angular3',
    }, {
        userId: 2,
        id: 4,
        body: 'Hello World4',
        title: 'testing Angular4',
    }],
    headers: {
        Link: '</api/v1/admin/stores/?sort=name&pageSize=20&next%5B%24or%5D%5B0%5D%5Bname%5D%5B%24gt%5D=LARI%20LUKE&next%5B%24or%5D%5B1%5D%5Bname%5D%5B%24gte%5D=LARI%20LUKE&next%5B%24or%5D%5B1%5D%5B_id%5D%5B%24gt%5D=5cffc130c82fe51c5afe0a9e>; rel="next"',
    },
};

const page3Data = {
    body: [{
        userId: 1,
        id: 5,
        body: 'Hello World3',
        title: 'testing Angular3',
    }],
    headers: {},
};

describe('ManyComponent', () => {
    let component: ManyComponent;
    let fixture: ComponentFixture<ManyComponent>;

    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [
                HttpClientTestingModule,
                NgxsModule.forRoot([EntityCrudEntitiesState, PostsEntitiesState, CommentsEntitiesState]),
                NgxsDataPluginModule.forRoot(),
                NgxsMiddlewareModule.forRoot(),
                ManyUiModule,
            ],
            providers: [
                {
                    provide: HTTP_INTERCEPTORS,
                    multi: true,
                    useClass: PaginationInterceptor,
                },
                TestPaginatedCrudService,
                TestCrudCollectionService,
            ],
        }).compileComponents();

        fixture = MockRender(`
            <frrri-many path="cache.posts">
                My content
                <div class="loading">mock-loading</div>
                <div class="loading-first">mock-first-load</div>
                <div class="load-more">mock-more</div>
            </frrri-many>
        `);

        component = fixture.debugElement
            .query(By.directive(ManyComponent))
            .componentInstance as ManyComponent;
    });

    it('should create and initialize', async () => {
        expect(component).toBeTruthy();
        const loading = await component.loading$.pipe(take(1)).toPromise();
        expect(loading).toBeFalsy();
    });

    it('should show contents correctly', inject([
        PostsEntitiesState,
        TestPaginatedCrudService,
    ], (
        postsEntities: PostsEntitiesState,
        service: TestPaginatedCrudService,
    ) => {
        // INIT
        expect(fixture.nativeElement.textContent.trim()).toEqual('My content');
        expect(fixture).toMatchSnapshot('init');
        const subject = new Subject();

        // PAGE 1/3
        spyOn(service, 'getMany').and.returnValue(subject.asObservable());
        postsEntities.getMany().toPromise();

        // PAGE 1/3 - LOADING FIRST
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot('loading-page-1-of-3');

        // // PAGE 1/3 - LOADED
        subject.next({ pagination: { data: page1Data.body, next: page1Data.headers.Link } });
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot('loaded-page-1-of-3');

        // // PAGE 2/3
        spyOn(service, 'getNext').and.returnValue(subject.asObservable());
        postsEntities.getNext().toPromise();

        fixture.detectChanges();

        // // PAGE 2/3 - LOADING
        expect(fixture).toMatchSnapshot('loading-page-2-of-3');
        subject.next({ pagination: { data: page2Data.body, next: page2Data.headers.Link } });
        fixture.detectChanges();

        // // PAGE 2/3 LOADED
        expect(fixture).toMatchSnapshot('loaded-page-2-of-3');

        // // PAGE 3/3 (LAST PAGE)
        postsEntities.getNext().toPromise();
        fixture.detectChanges();

        // // PAGE 3/3 LOADING
        expect(fixture).toMatchSnapshot('loading-page-3-of-3');
        subject.next({ pagination: { data: page3Data.body } });
        fixture.detectChanges();

        // // PAGE 3/3 LOADED
        expect(fixture).toMatchSnapshot('loaded-page-3-of-3');
    }));

    it('should show errors', inject([
        PostsEntitiesState,
        TestPaginatedCrudService,
    ], (
        postsEntities: PostsEntitiesState,
        service: TestPaginatedCrudService,
    ) => {
        // INIT
        expect(fixture.nativeElement.textContent.trim()).toEqual('My content');
        expect(fixture).toMatchSnapshot('init');
        const subject = new Subject();

        // PAGE 1/3
        spyOn(service, 'getMany').and.returnValue(subject.asObservable().pipe(
            tap((data) => {
                if (data === 'error') {
                    throw new Error('Http failure response for https://jsonplaceholder.typicode.com/posts: 400 Bad Request');
                }
            }),
        ));
        postsEntities.getMany().toPromise();
        fixture.detectChanges();

        // PAGE 1/3 - LOADING FIRST
        expect(fixture).toMatchSnapshot('loading-page-1-of-3');
        subject.next('error');
        fixture.detectChanges();

        // PAGE 1/3 - ERROR
        expect(fixture).toMatchSnapshot('error-page-1-of-3');
    }));

    afterEach(inject([HttpTestingController], (httpMock: HttpTestingController) => {
        httpMock.verify();
    }));
});
