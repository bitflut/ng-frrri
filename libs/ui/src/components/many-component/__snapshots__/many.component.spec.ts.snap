// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManyComponent should show contents correctly: init 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
  </frrri-many>
</mock-render>
`;

exports[`ManyComponent should show contents correctly: loaded-page-1-of-3 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
    <div
      class="load-more-content"
    >
      <div>
        <div>
          <div
            class="load-more"
          >
            mock-more
          </div>
        </div>
        
      </div>
    </div>
    
    
    
    
    
    
  </frrri-many>
</mock-render>
`;

exports[`ManyComponent should show contents correctly: loaded-page-2-of-3 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
    <div
      class="load-more-content"
    >
      <div>
        <div>
          <div
            class="load-more"
          >
            mock-more
          </div>
        </div>
        
      </div>
    </div>
    
    
    
    
    
    
  </frrri-many>
</mock-render>
`;

exports[`ManyComponent should show contents correctly: loaded-page-3-of-3 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
  </frrri-many>
</mock-render>
`;

exports[`ManyComponent should show contents correctly: loading-page-1-of-3 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    <div
      class="loading-first-content"
    >
      <div>
        <div
          class="loading-first"
        >
          mock-first-load
        </div>
      </div>
      
    </div>
    
    
    
    
    
  </frrri-many>
</mock-render>
`;

exports[`ManyComponent should show contents correctly: loading-page-2-of-3 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    <div
      class="loading-content"
    >
      <div>
        <div
          class="loading"
        >
          mock-loading
        </div>
      </div>
    </div>
    
    
    
    
    
    <div
      class="loading-more-content"
    >
      <div />
      
      <div>
         Loading more... 
      </div>
    </div>
    
    
    
    
    
    
  </frrri-many>
</mock-render>
`;

exports[`ManyComponent should show contents correctly: loading-page-3-of-3 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    <div
      class="loading-content"
    >
      <div>
        <div
          class="loading"
        >
          mock-loading
        </div>
      </div>
    </div>
    
    
    
    
    
    <div
      class="loading-more-content"
    >
      <div />
      
      <div>
         Loading more... 
      </div>
    </div>
    
    
    
    
    
    
  </frrri-many>
</mock-render>
`;

exports[`ManyComponent should show errors: error-page-1-of-3 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    <div
      class="error-content"
    >
      <div />
      
      
      <div
        style="color: red;"
      >
        Http failure response for https://jsonplaceholder.typicode.com/posts: 400 Bad Request
      </div>
    </div>
    
    
    
  </frrri-many>
</mock-render>
`;

exports[`ManyComponent should show errors: init 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    
    
    
    
    
    
    
  </frrri-many>
</mock-render>
`;

exports[`ManyComponent should show errors: loading-page-1-of-3 1`] = `
<mock-render>
  <frrri-many
    path="cache.posts"
  >
    <div
      class="content"
    >
       My content 
    </div>
    
    <div
      class="loading-first-content"
    >
      <div>
        <div
          class="loading-first"
        >
          mock-first-load
        </div>
      </div>
      
    </div>
    
    
    
    
    
  </frrri-many>
</mock-render>
`;
